lib/b3fs/src/bucket/file/writer.rs

```diff
@@ -1,29 +1,211 @@
+use crate::hasher::byte_hasher::Blake3Hasher;
+use crate::hasher::collector::BufCollector;
+use crate::utils::{self, tree_index};
 
 pub struct FileWriter {
-    hasher: HashTreeBuilder,
+    bucket: Bucket,
+    hasher: Blake3Hasher<BufCollector>,
+    temp_file_path: PathBuf,
+    block_files: Vec<PathBuf>,
+    current_block_file: Option<(usize, File)>,
+    header_file: BufWriter<File>,
+    count_block: usize,
 }
 
 impl FileWriter {
-    pub fn new(bucket: &Bucket) -> Self {
-        todo!()
+    pub async fn new(bucket: &Bucket) -> Result<Self, errors::WriteError> {
+        let mut wal_path = bucket.get_new_wal_path().to_path_buf();
+        let temp_file_path = wal_path.clone();
+        tokio::fs::create_dir_all(&wal_path).await?;
+        let random_file: [u8; 32] = random();
+        wal_path.push(utils::to_hex(&random_file).as_str());
+        let mut header_file = BufWriter::new(
+            tokio::fs::OpenOptions::new()
+                .create(true)
+                .append(true)
+                .open(wal_path)
+                .await?,
+        );
+        let version: u32 = 1;
+        let num_entries: u32 = 0;
+        header_file.write_u32_le(version).await?;
+        header_file.write_u32_le(num_entries).await?;
+        Ok(Self {
+            bucket: bucket.clone(),
+            hasher: Blake3Hasher::default(),
+            temp_file_path,
+            count_block: 0,
+            block_files: Vec::new(),
+            current_block_file: None,
+            header_file,
+        })
     }
 
     pub async fn write(&mut self, bytes: &[u8]) -> Result<(), errors::WriteError> {
-        todo!()
+        // Wrap the bytes in a BytesMut so we can split it into chunks.
+        let mut bytes_mut = BytesMut::from(bytes);
+        // Write the bytes to the current random block file incrementally or create a new block file
+        // if the current block file is full.
+        while bytes_mut.has_remaining() {
+            let want = cmp::min(BLOCK_SIZE_IN_CHUNKS, bytes_mut.len());
+            let mut bytes = bytes_mut.split_to(want);
+            if let Some((ref mut size, ref mut current_block_file)) = self.current_block_file {
+                if *size + bytes.len() > MAX_BLOCK_SIZE_IN_BYTES {
+                    let remaining = MAX_BLOCK_SIZE_IN_BYTES - *size;
+                    let block_remaining = bytes.split_to(remaining);
+                    current_block_file.write_all(&block_remaining).await?;
+                    current_block_file.flush().await?;
+                    let mut buffer = BytesMut::with_capacity(MAX_BLOCK_SIZE_IN_BYTES);
+                    current_block_file.read_buf(&mut buffer).await?;
+                    self.hasher.update(&buffer);
+                    let block_hash = self
+                        .hasher
+                        .tree
+                        .get_block_hash(self.count_block)
+                        .ok_or(errors::WriteError::BlockHashNotFound)?;
```

Comment:
the 'hasher' only has the block hash of the block if it knows that that block is finished, and it only knows that it is when there is at least one more byte given to the hasher after the block's data.

So for example, if the block size is `256KB` and we feed the hasher 256KB of data, it still wouldn't be able to give us the hash of that block, but as soon as we feed it at least one more byte, at that point the hasher can see that the previous block is finished and it finalizes it at that point.

Which means here this function would always fail since we are feed the entire block to the hasher (with any more bytes) and we are immediately asking for the hash of that block as soon as we do it. But at this point the hasher doesn't even know that it's ready to finalize that block just yet.

So the control flow here kind of needs a bit of overall change so that we 'finalize' (that is commit, rename, or anything like that) a block also as soon as we get the first byte of the next block and then we can go and mark the previous one complete and ask the hasher for that block hash.

```plain
- feed 256KB
- > block filled
- feed 1 byte
- > pass it to hasher
- > now hasher's hash tree contains the hash of that first block
- > grab that from the hash tree
- > do any file rename and stuff like that
- > flush the previous file to the disk for the 'filled' block
- > create the new file for the new block
- > write that 1 byte into the new file
```